### PR TITLE
Fix server not running with explicit hostname

### LINKED ISSUE #825: Uvicorn fails to resolve host names in version 0.12.2
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

Since version 0.12.2 it is not possible to use host names (including localhost) with `uvicorn.run` 

### To reproduce

```
import uvicorn

async def app(scope, receive, send):
    assert scope['type'] == 'http'

    await send({
        'type': 'http.response.start',
        'status': 200,
        'headers': [
            [b'content-type', b'text/plain'],
        ],
    })
    await send({
        'type': 'http.response.body',
        'body': b'Hello, world!',
    })

uvicorn.run(app, host='localhost', port=8083)
```

### Expected behavior

The application should start

### Actual behavior

Uvicorn throws an error

### Debugging material

```
.py thon/debugpy/launcher 61232 -- /Users/paulafernandez/Sandbox/test/test/start 
INFO:     Started server process [43171]
INFO:     Waiting for application startup.
INFO:     ASGI 'lifespan' protocol appears unsupported.
INFO:     Application startup complete.
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/paulafernandez/.vscode/extensions/ms-python.python-2020.9.114305/pythonFiles/lib/python/debugpy/__main__.py", line 45, in <module>
    cli.main()
  File "/Users/paulafernandez/.vscode/extensions/ms-python.python-2020.9.114305/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/Users/paulafernandez/.vscode/extensions/ms-python.python-2020.9.114305/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 267, in run_file
    runpy.run_path(options.target, run_name=compat.force_str("__main__"))
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/paulafernandez/Sandbox/test/test/start.py", line 18, in <module>
    uvicorn.run(
  File "/Users/paulafernandez/Library/Caches/pypoetry/virtualenvs/test-8kdwd4GS-py3.8/lib/python3.8/site-packages/uvicorn/main.py", line 391, in run
    server.run()
  File "/Users/paulafernandez/Library/Caches/pypoetry/virtualenvs/test-8kdwd4GS-py3.8/lib/python3.8/site-packages/uvicorn/main.py", line 419, in run
    loop.run_until_complete(self.serve(sockets=sockets))
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/Users/paulafernandez/Library/Caches/pypoetry/virtualenvs/test-8kdwd4GS-py3.8/lib/python3.8/site-packages/uvicorn/main.py", line 436, in serve
    await self.startup(sockets=sockets)
  File "/Users/paulafernandez/Library/Caches/pypoetry/virtualenvs/test-8kdwd4GS-py3.8/lib/python3.8/site-packages/uvicorn/main.py", line 526, in startup
    if isinstance(ip_address(config.host), IPv6Address):
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/ipaddress.py", line 53, in ip_address
    raise ValueError('%r does not appear to be an IPv4 or IPv6 address' %
ValueError: 'localhost' does not appear to be an IPv4 or IPv6 address
```

### Environment

- OS: macOS Catalina
- Python: 3.8.5
- Uvicorn version: 0.12.2
- The exact command you're running uvicorn with: `uvicorn.run(app, host='localhost', port=8083)`

### Additional context

The problem isn't present in earlier versions, including 0.12.1

### PR DESCRIPTION
**Pull request by euri10**
Fixes #825 

I'm not super happy with the try catch on the ip adresss but couldn't think of something else, so if someone has a better idea please share !
Not happy because I'm not sure it accounts for all potential use cases,

### CODE REVIEW

File: uvicorn/config.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ class _IPKind(Enum):
 
 
 def _get_server_start_message(
-    host_ip_version: _IPKind = _IPKind.IPv4,
+    host_ip_version: Optional[_IPKind] = None,
```
Comment (cdeler) [inline uvicorn/config.py]: @euri10 
Looks like I was wrong proposing `_IPKind` enum in https://github.com/encode/uvicorn/pull/803 . Looks like this function has 2 modes: either "to get a message for ipv6" and "to get a message for ipv4/hostname"

What do you think if we delete the `_IPKind` enum and change this function signature to 
```suggestion
    *, is_ipv6_message: bool = False
```
(and also change the function body & usage)

File: uvicorn/main.py [hunk 2]
```diff
@@ -523,10 +523,14 @@ def _share_socket(sock: socket) -> socket:
             if port == 0:
                 port = server.sockets[0].getsockname()[1]
             protocol_name = "https" if config.ssl else "http"
-            if isinstance(ip_address(config.host), IPv6Address):
-                message, color_message = _get_server_start_message(_IPKind.IPv6)
-            else:
-                message, color_message = _get_server_start_message(_IPKind.IPv4)
+            try:
+                addr = ip_address(config.host)
```
Comment (cdeler) [inline uvicorn/main.py:526]: Looks like you cannot omit this try-except block (or to replace it by regex)

Another approach to check if it's an ip or hostname is to:
```python
socket.getaddrinfo(config.host, 0, socket.AF_UNSPEC, socket.SOCK_STREAM, 0, socket.AI_NUMERICHOST)
```
But it's also should be wrapped by try-except  statement

File: uvicorn/config.py (hunk 1 above)
Comment (euri10) [reply uvicorn/config.py]: yup good idea, I didnt see it coming either :hamburger:

**Review by cdeler — APPROVED** [review]
Approved, nice fix :octocat:

**Comment by florimondmanca** [comment]
For traceability: marking this as a cleanup of #803.

**Pull request merged**